clippy_lints/src/matches/match_single_binding.rs

```diff
@@ -379,14 +359,42 @@ fn sugg_with_curlies<'a>(
 
     let mut indent = " ".repeat(indent_of(cx, ex.span).unwrap_or(0));
     let (mut cbrace_start, mut cbrace_end) = (String::new(), String::new());
-    if !expr_in_nested_block(cx, match_expr)
-        && ((needs_var_binding && is_var_binding_used_later) || expr_must_have_curlies(cx, match_expr))
-    {
+    let mut add_curlies = || {
         cbrace_end = format!("\n{indent}}}");
         // Fix body indent due to the closure
         indent = " ".repeat(indent_of(cx, bind_names).unwrap_or(0));
         cbrace_start = format!("{{\n{indent}");
         snippet_body = reindent_snippet_if_in_block(&snippet_body, !assignment_str.is_empty());
+    };
+
+    if !expr_in_nested_block(cx, match_expr) {
+        let mut parent = cx.tcx.parent_hir_node(match_expr.hir_id);
+        if let Node::Expr(Expr {
+            kind: ExprKind::Assign(..),
+            hir_id,
+            ..
+        }) = parent
+        {
+            parent = cx.tcx.parent_hir_node(*hir_id);
+        }
+        if let Node::Stmt(stmt) = parent {
+            parent = cx.tcx.parent_hir_node(stmt.hir_id);
+        }
+
+        match parent {
+            Node::Block(..)
+            | Node::Expr(Expr {
+                kind: ExprKind::Block(..) | ExprKind::ConstBlock(..),
+                ..
+            }) => {
+                if needs_var_binding && is_var_binding_used_later {
+                    add_curlies();
+                }
+            },
+            Node::Expr(..) | Node::AnonConst(..) => add_curlies(),
+            Node::Arm(arm) if let ExprKind::Match(..) = arm.body.kind => add_curlies(),
+            _ => {},
+        }
```

Comment:
The only cases this needs to special case are the parent being either `Node::Stmt` or `Node::Block`. Every other case needs to add braces.

`ExprKind::Block` and `ExprKind::ConstBlock` can only be the direct parent of `Node::Block` so they don't need to be handled specially.

Assignment always needs braces since the value of the arm still needs to be the result.